ml3d/torch/models/pointnet.py

```diff
@@ -0,0 +1,396 @@
+            gamma=cfg_pipeline.scheduler_gamma)
+        return optimizer, scheduler
+
+    # Todo: Parameters not consistent for base class and usage
+    def preprocess(self, data, attr):
+        return data
+
+    # Todo: Parameters not consistent for base class and usage
+    def transform(self, data, attr):
+        point = data['point'].copy()
+
+        # Sub/super-sample to get consistent point cloud size for batching
+        # Todo: How to run semantic segmentation inference on different point cloud sizes?
+        choice = np.random.choice(len(point), self.num_points, replace=True)
+        point = point[choice, :]
+
+        if self.task == 'classification':
+            # Normalize and center
+            if self.normalize:
+                point, _ = transforms.trans_normalize(
+                    point, feat=None, t_normalize={'method': 'unit_sphere'})
+
+            # Data augmentation for classification training
+            if self.augment and attr['split'] in ['train', 'training']:
+                # Paper: "[...] Gaussian noise with zero mean and 0.02 standard deviation."
+                if self.augment == 'gaussian':
+                    point = transforms.trans_augment(
+                        point, {
+                            'rotation_method': 'vertical',
+                            'noise_type': 'gaussian',
+                            'noise_level': 0.002
+                        })
+                # Following original implementation at
+                # https://github.com/charlesq34/pointnet/blob/master/train.py
+                elif self.augment == 'uniform':
+                    point = transforms.trans_augment(point, {
+                        'rotation_method': 'vertical',
+                        'noise_type': 'uniform_clip'
+                    })
+
+        if data.get('feat') is not None:
+            point = np.concatenate([point, data['feat'][choice].copy()], axis=1)
+
+        label = data['label'] if self.task == 'classification' else data[
+            'label'][choice]
+
+        return {'point': point.T, 'labels': label}
+
+    def inference_begin(self, data):
+        attr = {'split': 'test'}
+        self.inference_data = self.preprocess(data, attr)
+
+    def inference_preprocess(self):
+        attr = {'split': 'test'}
+        data = self.transform(self.inference_data, attr)
+        inputs = {'data': data, 'attr': attr}
+        return self.batcher.collate_fn([inputs])
+
+    def inference_end(self, inputs, results):
+        # Todo: Correct?
+        if self.task == 'classification':
+            inference_result = {
+                'predict_labels':
+                    np.array([results[0].argmax().detach().cpu().numpy()]),
+                'predict_scores':
+                    torch.nn.functional.softmax(results[0],
+                                                dim=1).detach().cpu().numpy()
+            }
+        else:
+            inference_result = {
```

Comment:
This is probably incorrect for segmentation, as we have to cover all of the points. Here, only a subsampled version gets covered.

Reply:
Oh, really? I though it was standard to only return predictions for points present in the input, i.e. if the input is subsampled, the output should be as well. Otherwise there needs to be some hand-crafted solution like interpolation to extend predicted labels to uncovered points. Is this desirable?

Reply:
The inference pipeline is expected to return exact same shape as that of input. Here, input is not subsampled, but the output is.
If you look at `inference_preprocess` method, you call `self.transform(...)` which will subsample the input.